pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go

```diff
@@ -1382,6 +1382,9 @@ func TestSelectBestCandidate(t *testing.T) {
 			candidates, _, _ := pe.DryRunPreemption(ctx, tt.pod, nodeInfos, nil, offset, numCandidates)
 			s := pe.SelectCandidate(candidates)
 			if s == nil || len(s.Name()) == 0 {
+				if len(tt.expected) != 0 {
+					t.Errorf("expect any node in %v, but got empty", tt.expected)
```

Comment:
alternatively, you can just use `Fatalf` here and remove the line of `return` below.

Reply:
Actually, if `len(tt.expected) == 0`, probably we should `return` without error. So, let's just keep it.

Reply:
Should we add a test that covers this?
Is it even possible that `SelectCandidate` returns empty?

Reply:
Ah, good point. I thought `len(s.Name()) == 0` can happen, but seems there is no scenario that should be tested but returns s.Name == "". 

Let me close this.

/close

Reply:
We should have `t.Fatal` to clearly signal that this shouldn't happen, instead of simply returning.

Reply:
Right. Here is the follow up: https://github.com/kubernetes/kubernetes/pull/115031

Reply:
We do have a case where `SelectCandidate` returns empty:
```
		{
			name:           "no candidates found",
			registerPlugin: st.RegisterPluginAsExtensions(noderesources.Name, nodeResourcesFitFunc, "Filter", "PreFilter"),
			nodeNames:      []string{"node1", "node2"},
			pod:            st.MakePod().Name("p").UID("p").Priority(lowPriority).Req(largeRes).Obj(),
			pods: []*v1.Pod{
				st.MakePod().Name("p1").UID("p1").Node("node1").Priority(highPriority).Req(largeRes).StartTime(epochTime).Obj(),
				st.MakePod().Name("p2").UID("p2").Node("node2").Priority(highPriority).Req(largeRes).StartTime(epochTime).Obj(),
			},
			expected: nil,
		},
```

So I think the right fix should be adding the boundary case.

Reply:
That's the case `DryRunPreemption` returns empty, right? I'm not sure it's worth adding that test case on this test, because this test focuses on `SelectCandidate` func, also actually if no candidate is returned from `findCandidates`, `SelectCandidate` isn't executed.
https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/preemption/preemption.go#L173

Reply:
But what if we had candidates after `findCandidates`, but got empty one after `callExtenders`. Then we'll still have empty candidates.

The point here is if this is an integration test, we should have the case for no candidates, if this is a unit test, covering every possible scenarios also seems convincible.

This seems like a combined unit test...

Reply:
> But what if we had candidates after findCandidates, but got empty one after callExtenders. Then we'll still have empty candidates.

That's a good point, how about adding `if len(candidates) == 0 {// return early //}` after the `callExtenders` func as we do after `findCandidates`?

Reply:
em... didn't see any significant improvements in this, prefer to only add a test case here.